src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProject.cs

```diff
@@ -1231,7 +1231,18 @@ public void RemoveFromWorkspace()
                 // references being converted to metadata references (or vice versa) and we might either
                 // miss stopping a file watcher or might end up double-stopping a file watcher.
                 remainingMetadataReferences = w.CurrentSolution.GetRequiredProject(Id).MetadataReferences;
-                w.OnProjectRemoved(Id);
+                _workspace.RemoveProjectFromTrackingMaps_NoLock(Id);
+
+                // If this is our last project, remove the entire solution so components that expect to see a
+                // SolutionRemoved actually get it
+                if (w.CurrentSolution.ProjectIds.Count == 1)
+                {
+                    w.OnSolutionRemoved();
+                }
+                else
+                {
+                    w.OnProjectRemoved(Id);
```

Comment:
Do we still always want to fire onremoved here?

Reply:
We don't need to, since a SolutionRemoved implies the removal of all of the projects. And that's actually consistent with what we do for the first project, which we fire just a SolutionAdded and that carries along the project with it.